src/Groups/matrices/iso_oscar_gap.jl

```diff
@@ -1,78 +1,110 @@
-   domain::FqNmodFiniteField      # Oscar field
+# T: type of the domain
+# S: element type of the domain (S == elem_type(T))
+struct GenRingIso{T, S} <: Map{T,GapObj,SetMap,GenRingIso}
+   domain::T      # Oscar field
    codomain::GapObj          # GAP field
    f                      # function from the Oscar field to the GAP field
    f_inv               # its inverse
 end
 
 # isomorphism from the Oscar matrix space to the GAP matrix space
-struct GenMatIso <: Map{FqNmodMatSpace,GapObj,SetMap,GenMatIso}
-   domain::FqNmodMatSpace        # Oscar matrix space
+# T: type of the domain
+# S: element type base ring (!) of the domain, that is, the entries of the
+#    matrices are of type S (T <: MatSpace{S})
+struct GenMatIso{T, S} <: Map{T,GapObj,SetMap,GenMatIso}
+   domain::T        # Oscar matrix space
    codomain::GapObj             # GAP matrix space
    fr::GenRingIso             # see type above
    f                         # function from the Oscar matrix space to the GAP matrix space
    f_inv                  # its inverse
 end
 
-
 ################################################################################
 #
 #  Ring isomorphism
 #
 ################################################################################
 
-# return the GAP element corresponding to the Oscar element x
-function elem_f(x::fq_nmod, B,d)
-   COEF = [Int64(coeff(x,i)) for i in 0:d-1]
-   x_gap = sum([COEF[i]*B[i] for i in 1:d])
-   return x_gap
-end
+# computes the isomorphism between the Oscar field F and the corresponding GAP field F_gap
+# the output has type GenRingIso (see above)
+
+function gen_ring_iso(F::T) where T <: Union{Nemo.GaloisField, Nemo.GaloisFmpzField}
+   p = characteristic(F)
+   G = GAP.Globals.GF(GAP.julia_to_gap(p))
+
+   f(x::Union{Nemo.gfp_elem, Nemo.gfp_fmpz_elem}) = GAP.julia_to_gap(lift(x))*GAP.Globals.One(G)
 
-# return the Oscar element corresponding to the GAP element x
-function elem_g(x::FFE, B, z, d)
-   L = Vector{Int}(GAP.Globals.List(GAP.Globals.Coefficients(B,x),GAP.Globals.IntFFE ))
-   return sum([L[i]*z^(i-1) for i in 1:d])
+   # For "small" primes x should be an FFE, but for bigger ones it's GAP_jll.Mptr (?)
+   finv(x) = F(GAP.gap_to_julia(GAP.Globals.IntFFE(x)))
+
+   return GenRingIso{typeof(F), elem_type(F)}(F, G, f, finv)
 end
 
+function gen_ring_iso(F::T) where T <: Union{Nemo.FqNmodFiniteField, Nemo.FqFiniteField}
+   p = characteristic(F)
+   d = degree(F)
+   Fp_gap = GAP.Globals.GF(GAP.julia_to_gap(p))
 
-# computes the isomorphism between the Oscar field F and the corresponding GAP field F_gap
-# the output has type GenRingIso (see above)
+   if d == 1
+      f1(x::Union{Nemo.fq_nmod, Nemo.fq}) = GAP.julia_to_gap(coeff(x, 0))*GAP.Globals.One(Fp_gap)
 
-function gen_ring_iso(F::FqNmodFiniteField)
-   p = Int64(characteristic(F))
-   d = Int64(degree(F))
-   z = gen(F)
+      # For "small" primes x should be an FFE, but for bigger ones it's GAP_jll.Mptr (?)
```

Comment:
FFE's only work for fields up to size 2^16, afterwards a different representation is used, which is a generic GAP object; on the Julia level, these all have the same type `GAP_jll.MPtr` (also known as `GAP.GapObj`; I should figure out a way to let Julia print it nicer)

Reply:
Ah, okay. I mainly added that comment to point out, why the function does not restrict it's input type.